### PR TITLE
Allow Extra Raster-Vision Segmentation Class to be Ignored

### PR DESCRIPTION
**Pull request by jamesmcclain**

### CODE REVIEW

File: focal_loss.py [hunk 1]
```diff
@@ -51,6 +53,13 @@ def forward(self, x: Tensor, y: Tensor) -> Tensor:
 
         # compute weighted cross entropy term: -alpha * log(pt)
         log_p = F.log_softmax(x, dim=-1)
+        if self.nll_loss.ignore_index == x.shape[1]:
```
Comment (jamesmcclain) [inline focal_loss.py:56]: "if the ignore index is one more than the index values produced by the mode ..."

File: focal_loss.py [hunk 2]
```diff
@@ -51,6 +53,13 @@ def forward(self, x: Tensor, y: Tensor) -> Tensor:
...
+            ignore_column = torch.tensor(log_p.shape[0] * [-math.inf],
```
Comment (jamesmcclain) [inline focal_loss.py:57]: Give the non-existent class zero probability

**Comment by AdeelH** [comment]
This is not necessary. The following runs without errors:
```
torch.nn.functional.nll_loss(torch.rand(10, 3), torch.full((10,), 3.).long(), ignore_index=3)
```

**Comment by jamesmcclain** [comment]
> This is not necessary. The following runs without errors:
> 
> ```
> torch.nn.functional.nll_loss(torch.rand(10, 3), torch.full((10,), 3.).long(), ignore_index=3)
> ```

Hmmm ... it seem to be empirically necessary when used with RV.  If you have a three-class problem, RV will add an extra class 4th class (making [this](https://github.com/azavea/noaa-flood-mapping/blob/master/pipeline/usfimr_raster_pipeline.py#L196)  necessary) even though the model itself only [only generates three classes](https://github.com/azavea/noaa-flood-mapping/blob/master/pipeline/usfimr_raster_pipeline.py#L224).

My best argument is to try to run the NOAA pipeline without [this](https://github.com/azavea/noaa-flood-mapping/blob/master/pipeline/usfimr_raster_pipeline.py#L207).

(BTW: I am already setting [`ignore_index`](https://github.com/azavea/noaa-flood-mapping/blob/master/pipeline/usfimr_raster_pipeline.py#L214) this change is to prevent [this line](https://github.com/AdeelH/pytorch-multi-class-focal-loss/blob/de74657769e07dc40be838a6277dea269bfddad0/focal_loss.py#L58) from raising an exception).

**Comment by AdeelH** [comment]
> (BTW: I am already setting ignore_index this change is to prevent this line from raising an exception).

Ah, that makes sense. There should be a simpler way to solve this though. I'll look into it.

**Comment by jamesmcclain** [comment]
> > (BTW: I am already setting ignore_index this change is to prevent this line from raising an exception).
> 
> Ah, that makes sense. There should be a simpler way to solve this though. I'll look into it.

:+1: I'm all for that.  If nothing is obvious, it might be better to put this behavior under a `raster_vision_workaround` argument to the entry point rather than doing it the way it is currently done.

**Comment by AdeelH** [comment]
Can you do a quick test of the fix in `AdeelH/pytorch-multi-class-focal-loss:ignore_index`?

**Comment by jamesmcclain** [comment]
> Can you do a quick test of the fix in `AdeelH/pytorch-multi-class-focal-loss:ignore_index`?

It seems to work :+1:

**Comment by AdeelH** [comment]
Bug fixed in #2

**Comment by jamesmcclain** [comment]
Please don't delete the branch `AdeelH/pytorch-multi-class-focal-loss:ignore_index` for a little while, I have jobs that are depending on it.
